docs/proposals/dynamic-kubelet-settings.md

```diff
@@ -0,0 +1,129 @@
+
+
+## Goals of this design:
+
+- Add a well-defined, versioned configmap object for Kubelet configuration.
+- Add/Remove/Update global and per-node Kubelet config on the fly via the API server.
+- Provide a reliable way to get the configuration of a given Kubelet along with its source (remote or on disk or flags).
+- We want the Kubelet to work with the built-in defaults on all distros we test against.
+- Work towards deprecating flags in favor of on-disk config and config from the API server. Master nodes will be bootstrapped from on-disk config.
+- Make it possible to opt-out of remote configuration e.g. to protect the master from bad config updates.
+
+## Design
+
+### Definitions
+
+- `kubelet-default`: The built-in default config that every Kubelet starts up with.
+- `api-default`: The defaults provided for omitted/nil fields by the API server.
+- `on-disk-config`: `KubeletConfiguration` provided on disk, used as a fallback when the Kubelet cannot contact the API server.
+- `minimal-config`: Conceptually, what the Kubelet needs to start up and contact the API server. This probably consists of the location of the API server and credentials to access the API server, and might be provided via flags or via `on-disk-config`.
+- `global-kubelet-configuration`: the `KubeletConfiguration` configmap set via the API server which is shared by all Kubelets in your cluster except those which have a `<node-name>-kubelet-configuration`.
+- `<node-name>-kubelet-configuration`: the `KubeletConfiguration` configmap specific to the node named `<node-name>`, also set via the API server.
+
+### Well-defined configmap type for Kubelet configuration
+
+- There is an alpha version of the `KubeletConfiguration` type: external in `pkg/componentconfig/v1alpha1/types.go`, internal in `pkg/componentconfig/types.go`.
+- We likely want to group related fields of the `KubeletConfiguration` into substructures. Please note that `v1alpha1` will not be regrouped, but eventually we will start working on this.
+
+@timstclair and I did some brainstorming and came up with some rough potential categories:
+
+---
+- Networking
+- Runtimes (only one of these, in addition to Common, should be specified for a given Kubelet)
+    + Common
+    + Rkt
+    + Docker
+- Events
+- Security/Policies
+- Images (might be better as a runtime sub-category)
+- Resource Management
+    + GC/Eviction
+    + Allocatable/SystemReserved/KubeReserved
+    + Custom resources? e.g. NvidiaGPUs?
+- Cloud-provider-specific settings
+- Cluster/API
+    + KubeAPIQPS
+    + ContentType
+- SystemSpecification (this might be a non-category, all of the available hardware should theoretically be detectable by the Kubelet)
+    + NodeIP (probably belongs in networking)
+    + NvidiaGPUs could be here, but shouldn't the Kubelet be able to detect this?
+- Experimental/Volatile Features (could be tricky, might be better off using alpha versioning to separate experiments from the stable API)
```

Comment:
This is related to #30003. Probably want something unversioned to toggle feature config.